File: python/OOP/pattern_factory.py
'''
Factory Pattern is used to create objects without exposing the instantiation logic

Useful when:
- the object creation process is complex
- we need to return a specific implementation of an interface

✅ Encapsulates object creation
✅ Supports multiple implementations
'''

class Animal:
    def speak(self):
        pass

class Dog(Animal):
    def speak(self):
        return "Woof!"

class Cat(Animal):
    def speak(self):
        return "Meow!"

class AnimalFactory:
    @staticmethod
    def create_animal(animal_type):
        if animal_type == "dog":
            return Dog()
        elif animal_type == "cat":
            return Cat()
        else:
            raise ValueError("Unknown animal type")

animal = AnimalFactory.create_animal("dog")
print(animal.speak())  # Output: "Woof!"

File: golang/07_application_structure_and_patterns/pattern_factory.go
package main
import "fmt"

/*
Factory Pattern is used to create objects without exposing the instantiation logic

Useful when:
- the object creation process is complex
- we need to return a specific implementation of an interface

✅ Encapsulates object creation
✅ Supports multiple implementations
*/

// Payment interface
type Payment interface {
    Pay(amount float64)
}

// CreditCard struct
type CreditCard struct{}

func (c CreditCard) Pay(amount float64) {
    fmt.Println("Paid", amount, "using Credit Card")
}

// PayPal struct
type PayPal struct{}

func (p PayPal) Pay(amount float64) {
    fmt.Println("Paid", amount, "using PayPal")
}

// Factory function
func NewPaymentMethod(method string) Payment {
    switch method {
    case "credit":
        return CreditCard{}
    case "paypal":
        return PayPal{}
    default:
        return nil
    }
}

func factory_pattern() {
    payment := NewPaymentMethod("credit")
    if payment != nil {
        payment.Pay(100)
    }
}
